vtk-m@2.0.0 +rocm build fails: vtkm/Atomic.h:334:60: error: 'memory_order_relaxed' is not a member of 'Kokkos::Impl'; did you mean 'std::memory _order_relaxed'?

### Steps to reproduce the issue

`vtk-m@2.0.0 +rocm` fails build using:
* spack@develop (7d072cc16f94abe9afc5066f739ff1175d9962bf from `Fri Sep 29 16:17:10 2023 -0400`)
* ROCm 5.4.3
* GCC 11.4.0
* Ubuntu 20.04, X86_64

Spack environment:  [spack.yaml.txt](https://github.com/spack/spack/files/12777561/spack.yaml.txt)

<details><summary>Concretization</summary>

<pre>
 -   27z2xiy  vtk-m@2.0.0%gcc@11.4.0~64bitids+ascent_types~cuda+doubleprecision+examples~fpic~ipo+kokkos~logging+mpi~openmp+rendering+rocm~shared~tbb~testlib amdgpu_target=gfx908 build_system=cmake build_type=Release generator=make arch=linux-ubuntu20.04-x86_64_v3
[+]  d6arsmf      ^cmake@3.27.6%gcc@11.4.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-ubuntu20.04-x86_64_v3
[+]  njq5f54          ^curl@8.1.2%gcc@11.4.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-ubuntu20.04-x86_64_v3
[+]  agoihmj              ^nghttp2@1.52.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]  tlxlem3              ^openssl@3.1.3%gcc@11.4.0~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu20.04-x86_64_v3
[+]  ty3g3w3                  ^ca-certificates-mozilla@2023-05-30%gcc@11.4.0 build_system=generic arch=linux-ubuntu20.04-x86_64_v3
[+]  nriy3cf                  ^perl@5.38.0%gcc@11.4.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-ubuntu20.04-x86_64_v3
[+]  6xoxed4                      ^berkeley-db@18.1.40%gcc@11.4.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu20.04-x86_64_v3
[+]  4ofiqoj          ^ncurses@6.3%gcc@11.4.0~symlinks+termlib abi=none build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]  x4n7p2a          ^zlib-ng@2.1.3%gcc@11.4.0+compat+opt build_system=autotools patches=299b958,ae9077a,b692621 arch=linux-ubuntu20.04-x86_64_v3
[+]  b7ftedl      ^gmake@4.4.1%gcc@11.4.0~guile build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[e]  hilucj5      ^hip@5.4.3%gcc@11.4.0~cuda+rocm build_system=cmake build_type=Release generator=make patches=5068750,c2ee21c,ca523f1,ddd86f0 arch=linux-ubuntu20.04-x86_64_v3
[e]  oiwskhu      ^hsa-rocr-dev@5.4.3%gcc@11.4.0+image+shared build_system=cmake build_type=Release generator=make patches=71e6851 arch=linux-ubuntu20.04-x86_64_v3
[+]  eynctgl      ^kokkos@4.1.00%gcc@11.4.0~aggressive_vectorization~compiler_warnings~cuda~debug~debug_bounds_check~debug_dualview_modify_check~deprecated_code~examples~hpx~hpx_async_dispatch~hwloc~ipo~memkind~numactl~openmp~openmptarget~pic+rocm+serial+shared~sycl~tests~threads~tuning~wrapper amdgpu_target=gfx908 build_system=cmake build_type=Release cxxstd=17 generator=make intel_gpu_arch=none arch=linux-ubuntu20.04-x86_64_v3
[e]  n3627rl      ^llvm-amdgpu@5.4.3%gcc@11.4.0~link_llvm_dylib~llvm_dylib~openmp+rocm-device-libs build_system=cmake build_type=Release generator=ninja patches=a08bbe1 arch=linux-ubuntu20.04-x86_64_v3
[+]  zrqfodp      ^mpich@4.1.2%gcc@11.4.0~argobots~cuda+fortran~hwloc+hydra+libxml2+pci~rocm+romio~slurm~vci~verbs~wrapperrpath build_system=autotools datatype-engine=auto device=ch4 netmod=ofi pmi=pmi arch=linux-ubuntu20.04-x86_64_v3
[+]  b2l3ksq          ^findutils@4.9.0%gcc@11.4.0 build_system=autotools patches=440b954 arch=linux-ubuntu20.04-x86_64_v3
[+]  mfz2wst          ^libfabric@1.19.0%gcc@11.4.0~debug~kdreg build_system=autotools fabrics=rxm,sockets,tcp,udp arch=linux-ubuntu20.04-x86_64_v3
[+]  rq4e3je          ^libpciaccess@0.17%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]  bvxwwuz              ^libtool@2.4.7%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]  3lzrbaf              ^util-macros@1.19.3%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]  ld7vpvc          ^libxml2@2.10.3%gcc@11.4.0+pic~python+shared build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]  6fs5xqm              ^libiconv@1.17%gcc@11.4.0 build_system=autotools libs=shared,static arch=linux-ubuntu20.04-x86_64_v3
[+]  gbz3mcq              ^xz@5.4.1%gcc@11.4.0+pic build_system=autotools libs=shared,static arch=linux-ubuntu20.04-x86_64_v3
[+]  rw7mj5e          ^pkgconf@1.9.5%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]  pewy6wy          ^yaksa@0.2%gcc@11.4.0~cuda~rocm build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]  klu7n45              ^autoconf@2.69%gcc@11.4.0 build_system=autotools patches=35c4492,7793209,a49dd5b arch=linux-ubuntu20.04-x86_64_v3
[+]  64zabhm              ^automake@1.16.5%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]  7gho7qw              ^m4@1.4.19%gcc@11.4.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-ubuntu20.04-x86_64_v3
[+]  uwo4zfi                  ^diffutils@3.9%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]  ffqewpm                  ^libsigsegv@2.14%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]  yo2ax24              ^python@3.11.4%gcc@11.4.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=13fa8bf,b0615b2,ebdca64,f2fd060 arch=linux-ubuntu20.04-x86_64_v3
[+]  yrn46p3                  ^bzip2@1.0.8%gcc@11.4.0~debug~pic+shared build_system=generic arch=linux-ubuntu20.04-x86_64_v3
[+]  iarwhi7                  ^expat@2.5.0%gcc@11.4.0+libbsd build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]  5aewl6n                      ^libbsd@0.11.7%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]  4nedds7                          ^libmd@1.0.4%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]  fvgp62q                  ^gdbm@1.23%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]  igbxcii                  ^gettext@0.21.1%gcc@11.4.0+bzip2+curses+git~libunistring+libxml2+tar+xz build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]  x5ok3wt                      ^tar@1.34%gcc@11.4.0 build_system=autotools zip=pigz arch=linux-ubuntu20.04-x86_64_v3
[+]  ahcrtma                          ^pigz@2.7%gcc@11.4.0 build_system=makefile arch=linux-ubuntu20.04-x86_64_v3
[+]  6cbxody                          ^zstd@1.5.5%gcc@11.4.0+programs build_system=makefile compression=none libs=shared,static arch=linux-ubuntu20.04-x86_64_v3
[+]  hxff7hm                  ^libffi@3.4.4%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]  lu5hko4                  ^libxcrypt@4.4.35%gcc@11.4.0~obsolete_api build_system=autotools patches=4885da3 arch=linux-ubuntu20.04-x86_64_v3
[+]  amkgvjk                  ^readline@8.2%gcc@11.4.0 build_system=autotools patches=bbf97f1 arch=linux-ubuntu20.04-x86_64_v3
[+]  vur4nkm                  ^sqlite@3.42.0%gcc@11.4.0+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]  3wnddk5                  ^util-linux-uuid@2.38.1%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
</pre></details>

Build error reproduced here using a Docker container image which includes ROCm 5.4.3 installed via `apt` to standard locations:
* `ecpe4s/ubuntu20.04-runner-amd64-gcc-11.4-rocm5.4.3:2023.08.01`
* Spack environment attached above includes the ROCm external definitions
```
$> spack -e . install
...
==> Installing vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w [19/19]
==> No binary for vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w found: installing from source
==> Using cached archive: /spack/var/spack/cache/_source-cache/archive/32/32643cf3564fa77f8e2a2a5456a574b6b2355bb68918eb62ccde493993ade1a3.tar.gz
==> No patches needed for vtk-m
==> vtk-m: Executing phase: 'cmake'
==> vtk-m: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    '/spack/opt/spack/linux-ubuntu20.04-x86_64_v3/gcc-11.4.0/gmake-4.4.1-b7ftedlmtybm64toojvb3fdbufhfm53l/bin/make' '-j16'

18 errors found in build log:
     143    [  8%] Building CXX object vtkm/cont/CMakeFiles/vtkm_cont.dir/CellLocatorUniformGrid.cxx.o
     144    cd /tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-build-27z2xiy/vtkm/cont && /spack/lib/spack/env/gcc/g++ -DVTKMDIY_MPI_AS_LIB -DVTKMDIY_MPI_STATIC_BUILD -DVTKMDIY_NO_THREADS -I/tmp/r
            oot/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-src/vtkm/thirdparty/optionparser -I/tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-src/vtkm/thirdparty/di
            y -I/tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-src/vtkm/thirdparty/lcl/vtkmlcl -I/tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-src -I/tmp/ro
            ot/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-build-27z2xiy/include -isystem /tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-src/vtkm/thirdparty/diy/vtk
            mdiy/include -isystem /tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-build-27z2xiy/vtkm/thirdparty/diy/vtkmdiy/include/vtkmdiy/mpi -isystem /spack/opt/spack/linux-ubuntu20.04-x86_64_v
            3/gcc-11.4.0/kokkos-4.1.00-eynctgljho6e53udsl4e2imu5bydtxtz/include -O3 -DNDEBUG -std=c++17 -fvisibility=hidden -Wall -Wcast-align -Wextra -Wpointer-arith -Wformat -Wformat-security -Wshadow -Wunused -fno-common -Wno-unus
            ed-function -Wchar-subscripts -Wfloat-conversion -Wodr -ffunction-sections -pthread -MD -MT vtkm/cont/CMakeFiles/vtkm_cont.dir/CellLocatorUniformGrid.cxx.o -MF CMakeFiles/vtkm_cont.dir/CellLocatorUniformGrid.cxx.o.d -o CM
            akeFiles/vtkm_cont.dir/CellLocatorUniformGrid.cxx.o -c /tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-src/vtkm/cont/CellLocatorUniformGrid.cxx
     145    cd /tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-build-27z2xiy/vtkm/cont && /spack/lib/spack/env/gcc/g++ -DVTKMDIY_MPI_AS_LIB -DVTKMDIY_MPI_STATIC_BUILD -DVTKMDIY_NO_THREADS -I/tmp/r
            oot/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-src/vtkm/thirdparty/optionparser -I/tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-src/vtkm/thirdparty/di
            y -I/tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-src/vtkm/thirdparty/lcl/vtkmlcl -I/tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-src -I/tmp/ro
            ot/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-build-27z2xiy/include -isystem /tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-src/vtkm/thirdparty/diy/vtk
            mdiy/include -isystem /tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-build-27z2xiy/vtkm/thirdparty/diy/vtkmdiy/include/vtkmdiy/mpi -isystem /spack/opt/spack/linux-ubuntu20.04-x86_64_v
            3/gcc-11.4.0/kokkos-4.1.00-eynctgljho6e53udsl4e2imu5bydtxtz/include -O3 -DNDEBUG -std=c++17 -fvisibility=hidden -Wall -Wcast-align -Wextra -Wpointer-arith -Wformat -Wformat-security -Wshadow -Wunused -fno-common -Wno-unus
            ed-function -Wchar-subscripts -Wfloat-conversion -Wodr -ffunction-sections -pthread -MD -MT vtkm/cont/CMakeFiles/vtkm_cont.dir/CellSet.cxx.o -MF CMakeFiles/vtkm_cont.dir/CellSet.cxx.o.d -o CMakeFiles/vtkm_cont.dir/CellSet
            .cxx.o -c /tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-src/vtkm/cont/CellSet.cxx
     146    In file included from /tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-src/vtkm/cont/BitField.h:17,
     147                     from /tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-src/vtkm/cont/BitField.cxx:11:
     148    /tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-src/vtkm/Atomic.h: In function 'T vtkm::detail::AtomicLoadImpl(T*, vtkm::MemoryOrder)':
  >> 149    /tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-src/vtkm/Atomic.h:334:60: error: 'memory_order_relaxed' is not a member of 'Kokkos::Impl'; did you mean 'std::memory_order_relaxed'?
     150      334 |       return Kokkos::Impl::atomic_load(addr, Kokkos::Impl::memory_order_relaxed);
     151          |                                                            ^~~~~~~~~~~~~~~~~~~~
     152    In file included from /usr/include/c++/11/bits/shared_ptr_atomic.h:33,
     153                     from /usr/include/c++/11/memory:78,
     154                     from /tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-src/vtkm/cont/Logging.h:21,
     155                     from /tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-src/vtkm/cont/Error.h:20,

     ...

     159                     from /tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-src/vtkm/cont/BitField.cxx:11:
     160    /usr/include/c++/11/bits/atomic_base.h:80:7: note: 'std::memory_order_relaxed' declared here
     161       80 |       memory_order_relaxed,
     162          |       ^~~~~~~~~~~~~~~~~~~~
     163    In file included from /tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-src/vtkm/cont/BitField.h:17,
     164                     from /tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-src/vtkm/cont/BitField.cxx:11:
  >> 165    /tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-src/vtkm/Atomic.h:338:60: error: 'memory_order_acquire' is not a member of 'Kokkos::Impl'; did you mean 'std::memory_order_acquire'?
     166      338 |       return Kokkos::Impl::atomic_load(addr, Kokkos::Impl::memory_order_acquire);
...
  >> 327    /tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-src/vtkm/Atomic.h:361:47: error: no member named 'memory_order_seq_cst' in namespace 'Kokkos::Impl'; did you mean '__memory_order_seq_cs
            t'?
     328          Kokkos::Impl::atomic_store(addr, value, Kokkos::Impl::memory_order_seq_cst);
     329                                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     330                                                  __memory_order_seq_cst
     331    /opt/rocm-5.4.3/include/hip/amd_detail/device_library_decls.h:146:3: note: '__memory_order_seq_cst' declared here
     332      __memory_order_seq_cst = __ATOMIC_SEQ_CST
     333      ^
     334    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-build-27z2xiy'
     335    [  8%] Built target vtkm_lodepng
     336    7 errors generated when compiling for gfx908.
  >> 337    make[2]: *** [vtkm/cont/CMakeFiles/vtkm_cont.dir/build.make:246: vtkm/cont/CMakeFiles/vtkm_cont.dir/CellLocatorUniformBins.cxx.o] Error 1
     338    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-build-27z2xiy'
  >> 339    make[1]: *** [CMakeFiles/Makefile2:2820: vtkm/cont/CMakeFiles/vtkm_cont.dir/all] Error 2
     340    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-vtk-m-2.0.0-27z2xiyevibfyygmbjdhadcncpx2i22w/spack-build-27z2xiy'
  >> 341    make: *** [Makefile:159: all] Error 2
```

### Error message

<details><summary>Error message</summary>

<pre>
... see above
</pre></details>


### Information on your system

* **Spack:** 0.21.0.dev0 (721398d95ed74c12824542203a6bcac66b5cd164)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-cascadelake
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/12777612/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/12777613/spack-build-out.txt)

@kmorel @vicentebolea @kwryankrattiger


### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/vtk-m/package.py
+++ b/var/spack/repos/builtin/packages/vtk-m/package.py
@@ -102,6 +102,7 @@ class VtkM(CMakePackage, CudaPackage, ROCmPackage):
 
     # VTK-m uses the default Kokkos backend
     depends_on("kokkos", when="+kokkos")
+    depends_on("kokkos@3.7:3.9", when="@2.0 +kokkos")
     # VTK-m native CUDA and Kokkos CUDA backends are not compatible
     depends_on("kokkos ~cuda", when="+kokkos +cuda +cuda_native")
     depends_on("kokkos +cuda", when="+kokkos +cuda ~cuda_native")

